### ISSUE TITLE
Pnpm errors when run with compatible "engines.pnpm" field in package.json

### ISSUE THREAD

**Issue by Siegrift** [pnpm/pnpm#7124]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

8.7.6

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Reproduction steps

This is part of my package.json:
```json
  "engines": {
    "node": "^18.14.0",
    "pnpm": "^8.7.6"
  }
```

When I run `pnpm run dev` I get:
```
 ERR_PNPM_UNSUPPORTED_ENGINE  Unsupported environment (bad pnpm and/or Node.js version)

Your pnpm version is incompatible with "<path-to-workspace>".

Expected version: ^8.7.6
Got: 8.7.5

This is happening because the package's manifest has an engines.pnpm field specified.
To fix this issue, install the required pnpm version globally.

To install the latest version of pnpm, run "pnpm i -g pnpm".
To check your pnpm version, run "pnpm -v".
 ELIFECYCLE  Command failed with exit code 1.
```

When I do `pnpm -v` I get:
```
8.7.6
```

### Describe the Bug

It seems the patch version of the CLI is not updated?

### Expected Behavior

The command should run without issues.

### Which Node.js version are you using?

18.14.0

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by kopach** [comment]
I have had multiple versions of `pnpm` installed. And for some reason in one case one version was used (e.g. when executed directly like `pnpm --version`) and another in different case (e.g. when run with `nx` in my case). To fix this, I had to uninstall `pnpm` with `pnpm un -g pnpm` and then running `which pnpm` or `pnpm --version` shown that another version still was available

**Comment by Siegrift** [comment]
Thanks, I actually suspected this and verified that `pnpm un -g pnpm` does nothing, but I didn't try using `which` and after these steps:

1. Add `"print-info": "which pnpm"` to package.json scripts
2. Run `pnpm run print-info`

I found out that indeed a different pnpm version was used, specifically: `/Users/<username>/.volta/tools/image/node/18.16.0/bin/pnpm` (instead of just `/Users/<username>/.volta/bin/pnpm`).

I managed to fix it by removing the image from volta cache and reinstalling. Specifically:

1. `volta install node@18.15` - change to different node version to "safely" remove
2. `rm -rf /Users/<username>/.volta/tools/image/node/18.16.0`
3. `pnpm run dev` - volta reinstalled the correct node version automatically and used correct (global) version of pnpm.
